src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs

```diff
@@ -0,0 +1,1745 @@
+#if TRAVE
+            if ((bytes[0] & 0x80) == 0)
+                // We have a three-byte header format
+                GlobalLog.Print("WARNING: SslState::DetectFraming() SSL v <=2 HELLO has no high bit set for 3 bytes header, we are broken, received byte = " + bytes[0].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
+
+            if (length < 3)
+            {
+                return Framing.Invalid;
+            }
+            if (bytes[2] > 8)
+                return Framing.Invalid;
+
+            if (bytes[2] == 0x1)  // SSL_MT_CLIENT_HELLO
+            {
+                if (length >= 5)
+                    version = (bytes[3] << 8) | bytes[4];
+            }
+            else if (bytes[2] == 0x4) // SSL_MT_SERVER_HELLO
+            {
+                if (length >= 7)
+                    version = (bytes[5] << 8) | bytes[6];
+            }
+
+            if (version != -1)
+            {
+                // If this is the first packet, the client may start with an SSL2 packet
+                // but stating that the version is 3.x, so check the full range.
+                // For the subsequent packets we assume that an SSL2 packet should have a 2.x version.
+                if (_Framing == Framing.None)
+                {
+                    if (version != 0x0002 && (version < 0x200 || version >= 0x500))
+                        return Framing.Invalid;
+                }
+                else
+                {
+                    if (version != 0x0002)
+                        return Framing.Invalid;
+                }
+            }
+
+            // When server has replied the framing is already fixed depending on the prior client packet
+            if (!Context.IsServer || _Framing == Framing.Unified)
+            {
+                return Framing.BeforeSSL3;
+            }
+
+            return Framing.Unified; // Will use Ssl2 just for this frame.
+        }
+
+        //
+        // This is called from SslStream class too.
+        internal int GetRemainingFrameSize(byte[] buffer, int dataSize)
+        {
+            GlobalLog.Enter("GetRemainingFrameSize", "dataSize = " + dataSize);
+            int payloadSize = -1;
+            switch (_Framing)
+            {
+                case Framing.Unified:
+                case Framing.BeforeSSL3:
+                    if (dataSize < 2)
+                    {
+                        throw new System.IO.IOException(SR.net_ssl_io_frame);
+                    }
+                    // Note: Cannot detect version mismatch for <= SSL2
+
+                    if ((buffer[0] & 0x80) != 0)
+                    {
+                        // Two bytes
+                        payloadSize = (((buffer[0] & 0x7f) << 8) | buffer[1]) + 2;
+                        payloadSize -= dataSize;
+                    }
+                    else
+                    {
+                        // Three bytes
+                        payloadSize = (((buffer[0] & 0x3f) << 8) | buffer[1]) + 3;
```

Comment:
Even if I understood the high bit test on line 1610, I wouldn't understand why you only consider 0x3f bytes in this case, but 0x7f in the previous one.  There's a lot of domain knowledge embedded in these magic numbers.